auth: Add option to restrict ldap login to realm by attribute

strifel opened the pull request:
This will provide an option to set an attribute that will restrict login to certain realms.

I use this since around a month now.
I wanted to have different realms to be accessible by different users (with a few common users), similar to #715.
I first tried this by having multiple custom AuthBackends but found myself having multiple issues
when it comes to synchronizing  user data from ldap.
I had no problems since I do it in the same way as this PR.
I wanted to ask if this has a chance of being merged, because I think this might help others in the same position, too.

**Testing Plan:** 
I tested it for around a month in prod now.
~I do have changed the name of the attribute in this PR.~ (Its now the same as I run myself)
I would be happy to implement tests if this PR has a chance of being accepted.

**possible improvements:**
- ~There might be users which want to check the realms against two LDAP attributes (e.g. department and office). So it might be a good Idea to accept a array of attributes~ (did that, because we do need it in our zulip installation)

strifel commented:
@mateuszmandera Thanks for having a look at this.

I saw that you changed the option to provide multiple attributes (back) to only one being accepted.
Can you please explain why you did it this way?
For us this would be a pretty important feature to match against multiple attributes as this would allow better integration into our existing LDAP schema. (e.g. have location and department realms)
If it is only to have a similar layout to SAML maybe the solution would be to allow both a list and a single attribute. (But than it has to be moved back to its own option)

mateuszmandera commented:
Hmm, I'm not sure what the right approach here is... Yeah I changed it for the sake of consistency  with SAML, and also I think in the general case it's rather unnatural to have multiple attributes with this purpose, although it suits your specific use case. Would it be problematic for you to adapt your ldap schema to have the extra attribute `zulip_org_membership` that would combine the information from the multiple attributes you're using currently?

strifel commented:
Last time I brought the idea into my group I was told that it would be better to change Zulip to accept multiple attributes (and I added it to this PR).
I do not know how unusual this is and I would think that there might relatively often be a use case for multiple attributes, but I do not really have examples.
I do think that not adding a special attribute would keep the LDAP cleaner, because changing one attribute would directly result in changing the access to Zulip.
But I definitely see you point, too.

mateuszmandera commented:
Hmm, yeahh not sure we should do here. I'm happy to adjust this to be a separate setting and allow multiple attributes if that will be more useful for our users despite being a bit less "clean" and natural. @timabbott What are your thoughts here?

timabbott commented:
Sorry for the delay!  I merged the two refactoring commits as the series ending with f078509105d9c71fa7e66be2c83862f982bacca7.

Do we need the `# nocoverage` in the last commit?  It seems like we have a test for that case.

In any case, I think we should integrate the `org_membership` field as the basic/generic use case, and add a second feature to cover @strifel's use case.  @strifel can you add the multi-fields version as a commit on top of this?

strifel commented:
@timabbott 
Thanks for reviewing!
Removed the `#nocoverage`
And added a "advanced LDAP realm access control" feature which should fit a lot of use cases

mateuszmandera commented:
I see one more problem with the first commit. Currently things are phrased in a way that talks about limit "access" based on the attribute - this won't be strictly true on deployments with additional auth backends enabled. Imagine:
1. A server has ldap and google backends enabled.
2. Originally the ldap-based restriction we're adding here is not enabled. john@zulip.com signs up via ldap to organization A.
3. The administrator decides "I don't want users such as john@zulip.com, who aren't explicitly permitted by an ldap attr, to access org A". Our docs talk about access, so the admin justifiably assumes it's enough to use the `org_membership` feature.
4. Now we have a problem, because john@zulip.com will indeed be disallowed by the ldap backend, but he can still log in with the Google backend.

I think we might need to document this more clearly?

strifel commented:
@mateuszmandera maybe it could be better called "allow access to subdomain by ldap authentication"?

As far as I can see [the SAML documentation has the same problem.](https://zulip.readthedocs.io/en/latest/production/authentication-methods.html#restricting-access-to-specific-organizations)

strifel commented:
@mateuszmandera I just cleared up the documentation in settings.py and
added documentation to the authentication-methods pages

mateuszmandera commented on docs/production/authentication-methods.md:
````diff
@@ -439,6 +439,37 @@ details.
+
+For example, with `org_membership` set to `department`, a user with
+the following attributes will have access to the root and `engineering` subdomains:
+```
+...
````
why do we have double ` ``` `?

strifel replied:
I do not really understand your question.
Line 456 are not ```, but dots to indicate that there are other attributes, that do not matter to us.

mateuszmandera replied:
Ah, github UI make the dots look like ````` to me :slightly_smiling_face: One more question, as I'm not that familiar with ldap syntax - is the example given here  how multivalued attributes are specificied? By having multiple lines with the attribute, specifying the values?

strifel replied:
I sadly could not find a source that shows that it is done like this.
Only thing I could find is [this Stackoverflow Awnser](https://stackoverflow.com/a/43155092/8314475).

mateuszmandera commented:
I assume we will want to somewhat adjust the phrasing when berging, but generally looks good. However, to make it possible to merge just the change in the first commit, could you split the doc changes and put them into the respective commits? (the first paragraph in the doc should go in the first commit, and the rest of it in the second commit.

The comments in `zproject/prod_settings_template.py` still has the old phrasing, so we'll want to fix it (perhaps just replace that commend with a reference to the doc to avoid duplication of explaining the same thing?).

> As far as I can see the SAML documentation has the same problem.

Hmm, could you point out which part of it could be understood as the restriction not being just SAML-specific? But I agree that at least we're not explicit about this subtlety in the doc, so perhaps we should add a warning similar to the one you're adding here?

mateuszmandera commented:
@strifel Cool, thanks for the changes! I made some tweaks to the documentation in the first commit and pushed here. Didn't touch the 2nd commit, as I assume Tim will have some more thoughts on how the advanced version of this feature should work in terms of configuration and logic.

@timabbott Can you review?

timabbott commented on zproject/prod_settings_template.py:
```diff
@@ -236,6 +236,11 @@
+    # This line is for having Zulip restricting user access to organizations to those
+    # listed in an LDAP attribute.
+    # See:
+    # https://zulip.readthedocs.io/en/latest/production/authentication-methods.html#restricting-ldap-user-access-to-specific-organizations
```
Can you shift the documentation to the `## ` style we now use in this file?

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -3782,6 +3782,40 @@ def test_ldap_auth_email_auth_disabled_success(self) -> None:
+        AUTHENTICATION_BACKENDS=("zproject.backends.ZulipLDAPAuthBackend",),
+        AUTH_LDAP_USER_ATTR_MAP={"full_name": "cn", "org_membership": "department"},
+    )
+    def test_ldap_auth_email_auth_organization_restriction_success(self) -> None:
```
I would actually delete the ~5 lines above this, combining the tests, and just remove the `_fail` from the name of the first test function.  It's often better to have shared test setup for failure/failure/success since then we know that the changes in the last version are what made it succeed.  (Currently the setups match, but one can more easily imagine that evolving with time if they're different functions).

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -3782,6 +3782,40 @@ def test_ldap_auth_email_auth_disabled_success(self) -> None:
+            "/api/v1/fetch_api_key",
+            dict(username=self.example_email("hamlet"), password=self.ldap_password("hamlet")),
+        )
+        self.assert_json_error(result, "Your username or password is incorrect.", 403)
```
Add a blank line after this, to visually separate the attempts, and move the "The second user ..." comment to after the blank line.

timabbott commented on docs/production/authentication-methods.md:
```diff
@@ -284,6 +284,34 @@ details.
+For the root subdomain, `www` in the list will work, or any other of
+`settings.ROOT_SUBDOMAIN_ALIASES`.
+
+For example, with `org_membership` set to `department`, a user with
```
How did we pick the `org_membership` field name?  That's the one remaining hangup I have about the non-documentation parts of this commit -- whether we're choosing the right name for this interface.

strifel replied:
I think @mateuszmandera set the name.
You mentioned it in https://github.com/zulip/zulip/pull/15659#issuecomment-696525019

As far as I understand we use this name to be consistent with the SAML options https://zulip.readthedocs.io/en/latest/production/authentication-methods.html#restricting-access-to-specific-organizations

mateuszmandera replied:
Yeah, my motivation was using the same name as in saml

timabbott commented on zproject/backends.py:
```diff
@@ -632,13 +632,32 @@ def is_account_control_disabled_user(self, ldap_user: _LDAPUser) -> bool:
+                        return False
+                # If no match is found forbid access
+                return True
+            else:
```
Does `AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL = {}` result in no enforcement?  

If so, we might want to have this always be a dictionary with a default of `{}`?

strifel replied:
I think currently `AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL = {}` would result in not being able to log in.
`AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL = None` and not having `org_membership` set, would result in no enforcement.

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -3816,6 +3816,55 @@ def test_ldap_auth_email_auth_organization_restriction_success(self) -> None:
+            "/api/v1/fetch_api_key",
+            dict(username=self.example_email("hamlet"), password=self.ldap_password("hamlet")),
+        )
+        self.assert_json_error(result, "Your username or password is incorrect.", 403)
```
I think my comments on the test setup/structure for the first commit apply to this second commit as well.

timabbott commented on zproject/backends.py:
```diff
@@ -632,13 +632,32 @@ def is_account_control_disabled_user(self, ldap_user: _LDAPUser) -> bool:
+            else:
+                # If configuration is wrong do not allow access
+                return True
+        else:
```
This doesn't need an `else` -- you can just `return False` after all the `if` blocks.

strifel replied:
I think it would not work here using just return `False` at the end of the method, because we return `True` in Line 657.

Because if at the end of the Method it checks if there are no settings set for `org_membership` or `ADVANCED_REALM_ACCESS_CONTROL` it does allow access.
In line 657 if a wrong configuration is made (e.g. not configuring a realm) we play safe and rather do not allow the access.

timabbott commented on zproject/backends.py:
```diff
@@ -632,13 +632,32 @@ def is_account_control_disabled_user(self, ldap_user: _LDAPUser) -> bool:
+        if "org_membership" in settings.AUTH_LDAP_USER_ATTR_MAP:
+            org_membership_attr = settings.AUTH_LDAP_USER_ATTR_MAP["org_membership"]
+            allowed_orgs: List[str] = ldap_user.attrs.get(org_membership_attr, [])
+            return not is_subdomain_in_allowed_subdomains_list(realm.subdomain, allowed_orgs)
```
If we want to make the two options not confusingly incompatible, which seems wise, I think this should be just 
```
if  not is_subdomain_in_allowed_subdomains_list(realm.subdomain, allowed_orgs):
    return True
```

and then fall through from there to check the second one.

timabbott commented on zproject/prod_settings_template.py:
```diff
@@ -243,6 +243,19 @@
+# This contains a map with the name of the Organizations (subdomain)
+# This contains a map with a attribute and a required value.
+# If one of these value is contained in the attribute access is granted
+# If nothing matches access through LDAP is not granted (but maybe still by different backends)
```
Let's move most of the documentation to ReadTheDocs, and have just an example here.

timabbott commented:
I merged the first commit as 209c89be100e8641821ffcc03bb8dac1b1f5edff after documentation tweaks, and pushed the second back after a bit more.  The main thing on my mind is that I think we're not leaving ourselves with the flexibility we'd ideally have for the `AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL` -- effectively what you have in an OR gate, with no natural way to add an AND gate as well.

There's two things I'd like to try:
* Make the values `AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL` into a list of dictionaries (as the OR gate), and enforce all of the properties.  E.g. `[{"department": "math"}, {"employeeType": "staff"}]` would be the new example for what you've done.
* The other extensibility issue is if we want to set properties (like role in the organization) based on these attributes, I'm not sure how we'd encode that.  One theory would be to use names like `require_department` and have the unprefixed namespace be usable for things like `role` if we later try to add that.

What do you think?  We can always migrate this without too much trouble since it's not likely to be used by tons of people for a while, but it seems worth thinking a bit about the structure before we merge.

strifel commented:
@timabbott Thanks.

The OR/AND idea is great and I implemented it.

I think setting the role is out of scope of the PR and would be done via a separate variable/setting, as one might not even want to use the same attributes, but doing different checks. (like checking if the user is an admin in AD).
I think `AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL` should just control access.

timabbott commented:
Yeah, that's reasonable -- I'm not saying we should try to build the `role` field, just thinking about how we'd extend this later if we thought it was a good idea.

timabbott commented:
I merged this with the following changes, mostly for readability (E.g. using the early return pattern for the check function).  Thanks for writing a thorough test suite, though next time I'd have used more declarative variable names.

```
diff --git a/zerver/tests/test_auth_backends.py b/zerver/tests/test_auth_backends.py
index 518e57f137..0baa458fbd 100644
--- a/zerver/tests/test_auth_backends.py
+++ b/zerver/tests/test_auth_backends.py
@@ -3837,6 +3837,7 @@ class FetchAPIKeyTest(ZulipTestCase):
             dict(username=self.example_email("hamlet"), password=self.ldap_password("hamlet")),
         )
         self.assert_json_error(result, "Your username or password is incorrect.", 403)
+
         self.change_ldap_user_attr("hamlet", "test1", "test")
         # Setting org_membership to not cause django_ldap_auth to warn, when synchronising
         self.change_ldap_user_attr("hamlet", "department", "wrongDepartment")
@@ -3865,7 +3866,7 @@ class FetchAPIKeyTest(ZulipTestCase):
             self.assert_json_success(result)
 
         # Setting test1 to wrong value
-        self.change_ldap_user_attr("hamlet", "test1", "test")
+        self.change_ldap_user_attr("hamlet", "test1", "invalid")
         result = self.client_post(
             "/api/v1/fetch_api_key",
             dict(username=self.example_email("hamlet"), password=self.ldap_password("hamlet")),
diff --git a/zproject/backends.py b/zproject/backends.py
index 266822e019..be442fbd4a 100644
--- a/zproject/backends.py
+++ b/zproject/backends.py
@@ -642,34 +642,36 @@ class ZulipLDAPAuthBackendBase(ZulipAuthMixin, LDAPBackend):
             if settings.AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL is None:
                 return True
 
+        # If neither setting is configured, allow access.
+        if settings.AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL is None:
+            return False
+
         # With settings.AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL, we
         # allow access if and only if one of the entries for the
         # target subdomain matches the user's LDAP attributes.
-        if settings.AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL is not None:
-            realm_access_control = settings.AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL
-            if (
-                isinstance(realm_access_control, dict)
-                and realm.subdomain in realm_access_control
-                and isinstance(realm_access_control[realm.subdomain], list)
-                and len(realm_access_control[realm.subdomain]) > 0
-            ):
-                # Go through every "or" check
-                for check in realm_access_control[realm.subdomain]:
-                    access = True
-                    for attribute in check:
-                        if not (
-                            attribute in ldap_user.attrs
-                            and check[attribute] in ldap_user.attrs[attribute]
-                        ):
-                            access = False
-                    if access:
-                        return False
-                # If no match is found forbid access
-                return True
-            # If configuration is wrong do not allow access
+        realm_access_control = settings.AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL
+        if not (
+            isinstance(realm_access_control, dict)
+            and realm.subdomain in realm_access_control
+            and isinstance(realm_access_control[realm.subdomain], list)
+            and len(realm_access_control[realm.subdomain]) > 0
+        ):
+            # If configuration is wrong, do not allow access
             return True
 
-        return False
+        # Go through every "or" check
+        for attribute_group in realm_access_control[realm.subdomain]:
+            access = True
+            for attribute in attribute_group:
+                if not (
+                    attribute in ldap_user.attrs
+                    and attribute_group[attribute] in ldap_user.attrs[attribute]
+                ):
+                    access = False
+            if access:
+                return False
+
+        return True
 
     @classmethod
     def get_mapped_name(cls, ldap_user: _LDAPUser) -> str:
```
